VitruvioHost/Plugins/Vitruvio/Source/VitruvioEditor/Private/VitruvioCooker.cpp

```diff
@@ -0,0 +1,332 @@
+
+			FMaterialCache MaterialCache;
+			FTextureCache TextureCache;
+			FStaticMeshCache MeshCache;
+
+			for (AActor* Actor : Actors)
+			{
+				CookTask.EnterProgressFrame(1);
+
+				AActor* OldAttachParent = Actor->GetAttachParentActor();
+
+				// Spawn new Actor with persisted geometry
+				AActor* CookedActor = Actor->GetWorld()->SpawnActor<AActor>(Actor->GetActorLocation(), Actor->GetActorRotation());
+				if (OldAttachParent)
+				{
+					CookedActor->AttachToActor(OldAttachParent, FAttachmentTransformRules::KeepWorldTransform);
+				}
+
+				USceneComponent* RootComponent = NewObject<USceneComponent>(CookedActor, "Root");
+				CookedActor->AddOwnedComponent(RootComponent);
+				CookedActor->SetRootComponent(RootComponent);
+				RootComponent->SetMobility(EComponentMobility::Movable);
+				RootComponent->OnComponentCreated();
+				RootComponent->SetWorldRotation(Actor->GetActorRotation());
+				RootComponent->SetWorldLocation(Actor->GetActorLocation());
+				RootComponent->RegisterComponent();
+
+				// Persist Mesh
+				UGeneratedModelStaticMeshComponent* StaticMeshComponent = Actor->FindComponentByClass<UGeneratedModelStaticMeshComponent>();
+				if (StaticMeshComponent)
+				{
+					check(StaticMeshComponent->GetStaticMesh());
+
+					UStaticMesh* GeneratedMesh = StaticMeshComponent->GetStaticMesh();
+
+					UStaticMesh* PersistedMesh = SaveStaticMesh(GeneratedMesh, CookPath, MeshCache, MaterialCache, TextureCache);
+					AttachMeshComponent<UStaticMeshComponent>(CookedActor, PersistedMesh, TEXT("Model"),
+															  StaticMeshComponent->GetComponentTransform());
+				}
+
+				// Persist instanced Component
+				TArray<UActorComponent*> HismComponents;
+				Actor->GetComponents(UGeneratedModelHISMComponent::StaticClass(), HismComponents);
+				for (UActorComponent* HismComponent : HismComponents)
+				{
+					UGeneratedModelHISMComponent* GeneratedModelHismComponent = Cast<UGeneratedModelHISMComponent>(HismComponent);
+					UStaticMesh* GeneratedMesh = GeneratedModelHismComponent->GetStaticMesh();
+					if (GeneratedMesh)
+					{
+						UStaticMesh* PersistedMesh = SaveStaticMesh(GeneratedMesh, CookPath, MeshCache, MaterialCache, TextureCache);
+
+						FName Name =
+							MakeUniqueObjectName(CookedActor, UHierarchicalInstancedStaticMeshComponent::StaticClass(), *PersistedMesh->GetName());
+						auto* InstancedStaticMeshComponent = AttachMeshComponent<UHierarchicalInstancedStaticMeshComponent>(
+							CookedActor, PersistedMesh, Name, GeneratedModelHismComponent->GetComponentTransform());
+
+						for (int32 InstanceIndex = 0; InstanceIndex < GeneratedModelHismComponent->GetInstanceCount(); ++InstanceIndex)
+						{
+							FTransform Transform;
+							GeneratedModelHismComponent->GetInstanceTransform(InstanceIndex, Transform);
+							InstancedStaticMeshComponent->AddInstance(Transform);
+						}
+					}
+				}
+
+				// Destroy the old procedural Vitruvio Actor
+				Actor->Destroy();
+			}
+
+			IsCooking = false;

```

Comment:
Probably doesn't matter, but I would move this line below `ModelsGeneratedHandle.Reset();`